docs/advanced.rst

```diff
@@ -243,3 +241,170 @@ the functions defined in their class body.
 functionality!
+
+.. _message-parts:
+
+Message Parts
+-------------
+In some advanced use cases, :attr:`.ChatMessage.content` may be a tuple of :class:`.MessagePart` or ``str`` rather than
+a string. ChatMessage exposes :attr:`.ChatMessage.text` (always a string or None) and :attr:`.ChatMessage.parts` (always
+a list of message parts), which we recommend using instead of :attr:`.ChatMessage.content`. These properties are
+dynamically generated based on the underlying content, and it is safe to mix messages with different content types in a
+single Kani.
+
+Generally, message part classes are *defined* by the engine, and *consumed* by the developer. Message parts can be used
+in any role's message - for example, you might use a message part in an assistant message to separate out a chain of
+thought from a user reply, or in a user message to supply an image to a multimodal model.
+
+Let's say you wanted your model to perform a chain-of-thought before replying to the user, but didn't want that
+chain of thought to be visible in its output. In the example below, we'll define a hidden *ThoughtPart* that contains
+this data, and have our engine use this when building the prompt. This may be a little bit contrived, but hopefully
+it demonstrates how to use the MessagePart interface.
+
+Defining a MessagePart
+^^^^^^^^^^^^^^^^^^^^^^
+To define a :class:`.MessagePart`, you should create a new class that inherits from :class:`.MessagePart`.
+
+Under the hood, a MessagePart is a Pydantic model, so to add attributes to your part, you can define them as
+type-annotated class variables. For example, to add a ``data`` attribute with a string type, you can define
+``data: str``.
+
+Next, you'll want to define the ``__str__()`` method, which is how kani defines the behaviour when your message part
+is cast to a string. This happens in a couple places:
+
+1. The message part is displayed to the user through a string-casting method (e.g. :meth:`.Kani.full_round_str`)
+2. The user is using the :func:`.chat_in_terminal` development utility
+3. The message was provided to an engine that does not support the class (e.g. an image part in a text-only model).
+
+The default behaviour is to transform the part to a Python-esque representation of its data and log a warning. You can
+override this method to specify the canonical string representation of your message part.
+
+.. code-block:: python
+
+    from kani import MessagePart
+
+    class ThoughtPart(MessagePart):
+        # hold some string data that is the chain-of-thought
+        data: str
+
+        # when stringified, this part should be hidden from the user/any engine that does not explicitly support it
+        def __str__(self):
+            return ""
+
+
+This will automatically register serialization and deserialization schemes for your class, so that all MessageParts
+are compatible with :meth:`.Kani.save` and :meth:`.Kani.load`.
```

Comment:
If you don't define the `__str__` method does that mean the serialization and deserialization breaks?